### ISSUE TITLE
Node not recovering from commit state after all the other nodes made a ChangeView

### ISSUE THREAD

**Issue by roman-khimov** [nspcc-dev/dbft#17]
Node number three is stuck in view 0 of block 6011 because it managed to reach a commit state:
```
2020-01-30T15:34:15.384Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:104     initialize      {"height": 6011, "view": 0, "index": 2, "role": "Backup"}
2020-01-30T15:34:15.384Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:549     reset timer     {"h": 6011, "v": 0, "delay": "10s"}
2020-01-30T15:34:15.628Z        INFO    blockchain persist completed    {"persistedBlocks": 1, "persistedKeys": 17393, "headerHeight": 6010, "blockHeight": 6010, "took": "38.976723ms"}
2020-01-30T15:34:16.183Z        DEBUG   consensus/consensus.go:177      received message        {"from": 3}
2020-01-30T15:34:16.183Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:203     received message        {"type": "PrepareResponse", "from": 3, "height": 6010, "view": 0, "my_height": 6011, "my_view": 0}
2020-01-30T15:34:16.183Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:212     ignoring old height     {"height": 6010}
2020-01-30T15:34:16.278Z        DEBUG   consensus/consensus.go:177      received message        {"from": 3}
2020-01-30T15:34:16.278Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:203     received message        {"type": "Commit", "from": 3, "height": 6010, "view": 0, "my_height": 6011, "my_view": 0}
2020-01-30T15:34:16.278Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:212     ignoring old height     {"height": 6010}
2020-01-30T15:34:24.823Z        DEBUG   consensus/consensus.go:177      received message        {"from": 3}
2020-01-30T15:34:24.823Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:203     received message        {"type": "PrepareRequest", "from": 3, "height": 6011, "view": 0, "my_height": 6011, "my_view": 0}
2020-01-30T15:34:26.651Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/send.go:9       broadcasting message    {"type": "PrepareResponse", "height": 6011, "view": 0}
2020-01-30T15:34:26.653Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/check.go:29     check preparations      {"hasReq": true, "count": 2, "M": 3}
2020-01-30T15:34:26.653Z        DEBUG   consensus/consensus.go:177      received message        {"from": 0}
2020-01-30T15:34:26.653Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:203     received message        {"type": "ChangeView", "from": 0, "height": 6011, "view": 0, "my_height": 6011, "my_view": 0}
2020-01-30T15:34:27.516Z        DEBUG   consensus/consensus.go:177      received message        {"from": 1}
2020-01-30T15:34:27.516Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:203     received message        {"type": "ChangeView", "from": 1, "height": 6011, "view": 0, "my_height": 6011, "my_view": 0}
2020-01-30T15:34:27.628Z        DEBUG   consensus/consensus.go:177      received message        {"from": 1}
2020-01-30T15:34:27.628Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:203     received message        {"type": "PrepareResponse", "from": 1, "height": 6011, "view": 0, "my_height": 6011, "my_view": 0}
2020-01-30T15:34:27.628Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:390     prepare response
2020-01-30T15:34:27.644Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/check.go:29     check preparations      {"hasReq": true, "count": 3, "M": 3}
2020-01-30T15:34:27.644Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/send.go:9       broadcasting message    {"type": "Commit", "height": 6011, "view": 0}
2020-01-30T15:34:27.645Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/check.go:57     not enough to commit    {"count": 1}
2020-01-30T15:34:27.678Z        DEBUG   consensus/consensus.go:177      received message        {"from": 0}
2020-01-30T15:34:27.678Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:203     received message        {"type": "PrepareResponse", "from": 0, "height": 6011, "view": 0, "my_height": 6011, "my_view": 0}
2020-01-30T15:34:27.678Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:390     prepare response
2020-01-30T15:34:32.644Z        DEBUG   consensus/consensus.go:172      timer fired     {"height": 6011, "view": 0}
2020-01-30T15:34:32.644Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:174     timeout {"height": 6011, "view": 0}
2020-01-30T15:34:32.644Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:182     send recovery to resend commit
2020-01-30T15:34:32.644Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/send.go:9       broadcasting message    {"type": "RecoveryRequest", "height": 6011, "view": 0}
2020-01-30T15:34:32.644Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:549     reset timer     {"h": 6011, "v": 0, "delay": "10s"}
2020-01-30T15:34:36.076Z        DEBUG   consensus/consensus.go:177      received message        {"from": 3}
2020-01-30T15:34:36.076Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:203     received message        {"type": "ChangeView", "from": 3, "height": 6011, "view": 0, "my_height": 6011, "my_view": 0}
2020-01-30T15:34:36.076Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:429     ignoring ChangeView: commit sent
2020-01-30T15:34:36.079Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/send.go:9       broadcasting message    {"type": "RecoveryMessage", "height": 6011, "view": 0}
2020-01-30T15:34:42.645Z        DEBUG   consensus/consensus.go:172      timer fired     {"height": 6011, "view": 0}
2020-01-30T15:34:42.645Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:174     timeout {"height": 6011, "view": 0}
2020-01-30T15:34:42.645Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:182     send recovery to resend commit
2020-01-30T15:34:42.645Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/send.go:9       broadcasting message    {"type": "RecoveryRequest", "height": 6011, "view": 0}
2020-01-30T15:34:42.645Z        DEBUG   dbft@v0.0.0-20200130105505-02c208d154bf/dbft.go:549     reset timer     {"h": 6011, "v": 0, "delay": "10s"}
2020-01-30T15:34:52.645Z        DEBUG   consensus/consensus.go:172      timer fired     {"height": 6011, "view": 0}
```

Maybe the ChangeView processing should be altered in that when the node sees that an M number of nodes announce a ChangeView it should follow the suit even if had managed to reach Commit phase, because those guys with changed view will not recover using the old one.

[four.log](https://github.com/nspcc-dev/dbft/files/4135677/four.log)
[one.log](https://github.com/nspcc-dev/dbft/files/4135678/one.log)
[three.log](https://github.com/nspcc-dev/dbft/files/4135679/three.log)
[two.log](https://github.com/nspcc-dev/dbft/files/4135680/two.log)

**Comment by roman-khimov** [comment]
Same thing here, the node number two is stuck in 6014/0.
[four.log](https://github.com/nspcc-dev/dbft/files/4138794/four.log)
[one.log](https://github.com/nspcc-dev/dbft/files/4138795/one.log)
[three.log](https://github.com/nspcc-dev/dbft/files/4138796/three.log)
[two.log](https://github.com/nspcc-dev/dbft/files/4138797/two.log)

**Comment by roman-khimov** [comment]
And even though it had received a new block the dBFT state was not reset!

**Comment by fyrchik** [comment]
Fixed in #22 .
